charts/ColorLegendView.tsx

```diff
@@ -0,0 +1,247 @@
+import { computed, action } from "mobx"
+import { observer } from "mobx-react"
+
+import { Bounds } from "./Bounds"
+import { getRelativeMouse, sortBy } from "./Util"
+import { ColorScaleBin, CategoricalBin } from "./ColorScaleBin"
+import {
+    ColorLegend,
+    NumericColorLegend,
+    CategoricalColorLegend
+} from "./ColorLegend"
+
+const FOCUS_BORDER_COLOR = "#111"
+
+export interface ColorLegendViewProps {
+    legend: ColorLegend
+    onMouseOver: (d: ColorScaleBin) => void
+    onMouseLeave: () => void
+}
+
+@observer
+export class ColorLegendView extends React.Component<ColorLegendViewProps> {
+    render() {
+        const { legend, onMouseOver, onMouseLeave } = this.props
+        const {
+            bounds,
+            mainLabel,
+            numericLegend,
+            categoryLegend,
+            categoryLegendHeight
+        } = legend
+
+        return (
+            <g className="mapLegend">
+                {numericLegend && (
+                    <NumericColorLegendView
+                        legend={numericLegend}
+                        x={bounds.centerX - numericLegend.width / 2}
+                        y={
+                            bounds.bottom -
+                            mainLabel.height -
+                            categoryLegendHeight -
+                            numericLegend.height -
+                            4
+                        }
+                        onMouseOver={onMouseOver}
+                        onMouseLeave={onMouseLeave}
+                    />
+                )}
+                {categoryLegend && (
+                    <CategoricalColorLegendView
+                        legend={categoryLegend}
+                        x={bounds.centerX - categoryLegend.width / 2}
+                        y={
+                            bounds.bottom -
+                            mainLabel.height -
+                            categoryLegendHeight
+                        }
+                        onMouseOver={onMouseOver}
+                        onMouseLeave={onMouseLeave}
+                    />
+                )}
+                {mainLabel.render(
+                    bounds.centerX - mainLabel.width / 2,
+                    bounds.bottom - mainLabel.height
+                )}
+            </g>
+        )
+    }
+}
+
+@observer
+class NumericColorLegendView extends React.Component<{
+    legend: NumericColorLegend
+    x: number
+    y: number
+    onMouseOver: (d: ColorScaleBin) => void
+    onMouseLeave: () => void
+}> {
+    base: React.RefObject<SVGGElement> = React.createRef()
+
+    @computed get bounds(): Bounds {
+        const { props } = this
+        return new Bounds(
+            props.x,
+            props.y,
+            props.legend.width,
+            props.legend.height
+        )
+    }
+
+    @computed get legend(): NumericColorLegend {
+        return this.props.legend
+    }
+
+    @action.bound onMouseMove(ev: MouseEvent | TouchEvent) {
```

Comment:
nit: might be helpful to add some explanatory comments in this method. I'm not 100% what is going on.

Reply:
I have not even looked at this, it's just code that's unchanged from before, just moved to a different file.